### DISCUSSION TITLE
Discourage use of `truncated_range` and `truncated_discrete_set` validators

### DISCUSSION THREAD

**Discussion by mcdo0486** [pymeasure/pymeasure#716] [Ideas]
Validators are typically used to validate the input value a user sets to an instrument property.

[truncated_range(value, values)](https://github.com/pymeasure/pymeasure/blob/9f50e169fa62bb4bbfa1ab0256045a314bfb6e59/pymeasure/instruments/validators.py#L84) and [truncated_discrete_set(value, values)](https://github.com/pymeasure/pymeasure/blob/9f50e169fa62bb4bbfa1ab0256045a314bfb6e59/pymeasure/instruments/validators.py#L125) are two validators that take in any numeric input as a `value` and returns a truncated output value that matches an element from the passed list of `values`.

Here are some examples from the [tests.instruments.test_validators](https://github.com/pymeasure/pymeasure/blob/9f50e169fa62bb4bbfa1ab0256045a314bfb6e59/tests/instruments/test_validators.py#L63):
```
truncated_range(20, range(10)) == 9
truncated_discrete_set(5.1, range(10)) == 6
truncated_discrete_set(11, range(10)) == 9
```

The point of these two validators is to make sure a valid, operational value is passed to the instrument during a procedure even if the input value is "out of bounds".

However, this default behavior for an instrument property is invisible to the user and can create significant problems.

First, incorrect parameters will be recorded in the results file.

A common procedure pattern is to include `FloatParameter` or `IntegerParameter` as inputs for an instrument. The results file will helpfully record these parameters along with the output data. If a user passes an “out of bounds” value that gets invisibly truncated up/down, it will be the “out of bounds” value that is recorded in the results file. That can create problems when output data is analyzed, because unless the person reading the results file happens to know their input parameter was truncated, they would assume the wrong value was used to generate the output data.

Second, procedures may be adapted or moved to other instruments where truncated behavior may be harmful.

The person that writes the original procedure may not be the one that always uses it. Procedures may be modified by others, or used with other instruments of near model/make. In a naive scenario, the original procedure may have used `inst.output_voltage = 2000` which was truncated down to 20 V without the author knowing it. If someone were to use an instrument that could output 2000 V, then that person may have a very bad day.

While the liability for coding or executing procedures is always on the person that is running the code, needlessly confusing or harmful default behaviors should be avoided if possible. Default behavior of validators should be explicit to the user with what value is passed to the instrument, and should `ValueError` out if the value is "out of bounds". 

`strict_discrete_set` and `strict_discrete_range` should be encouraged as default validators in documentation, and `truncated_range` and `truncated_discrete_set` should be discouraged for most use cases.

I also think usage of `truncated_range` and `truncated_discrete_set` should be reviewed in current instruments. Currently 31 instruments use `truncated_range` and 16 instruments use `truncated_discrete_set`. I’d be happy to help migrate instruments that I have access to.

**Comment by bilderbuchi** [comment]
Thank you for your thoughts about this! 
Personally, I have only ever found use for `truncated_range` (to clamp an input to a valid range), sometime its `strict_range` variant, and `strict_discrete_set` (to check if an entry is in an admissible set of values, e.g. string modes). 
Possibly the validators could use an overhaul in general, and a more composable, structured, and semantically more powerful set of validators. 

I definitely see your point about recording wrong values in a result file, that is a problem. Probably this could be solved by validators emitting a log message when they modify a value before passing, which the procedure machinery could intercept. @CasperSchippers, any input on that?

I agree that we should avoid needlessly confusing or harmful default behaviors if possible. 
However, I'd posit that the validators are actually safety features, to ensure that a harmful input cannot reach *a device*, where it could cause either some instrument error (overrange etc), throw the instrument into a bad state, or something else. The main difference between the strict and non-strict variant is if they except on you to signal that they had to act. 
Typically, I imagine you do not want, e.g., your multihour lab run to abort because your PID algorithm overreached and prescribed a PSU voltage of 101.2 V, when the PSU can only do 100, hence the usefulness of the non-strict variant.

> Default behavior of validators should be explicit to the user with what value is passed to the instrument, and should ValueError out if the value is "out of bounds".

I think that ultimately comes down to a personal preference. You prefer your procedure/script to except on you if any of the possibly dozens of parameters goes out of range (and probably to write a lot of recovery logic, or just start over). Other people prefer their runs to continue if feasible.
As I said above, that we don't have a feedback mechanism that a validator had to change inputs is something that probably should be changed.

Regarding the explicitness, this is ultimately a documentation/docstring issue, and we have some ideas about how to automate docstring creation so that the validator config becomes visible to a user (in the property's docstring) - see #571.

> strict_discrete_set and strict_discrete_range should be encouraged as default validators in documentation,

I'm curious, why do you prefer `strict_discrete_range` over `strict_range`?

**Reply by CasperSchippers** [reply]
I think it is a good idea to log a message/warning when a value is modified by a validator.

I do want to note that I think it is worth querying the actual value, if possible. I do of course see that this could create a lot of unnecessary overhead in some cases; for this #503 could be useful.

**Reply by bilderbuchi** [reply]
I filed #729 as a starting point.

**Reply by mcdo0486** [reply]
Good catch, I meant to include `strict_range` in that sentence, so "`strict_range`, `strict_discrete_set` and, `strict_discrete_range` should be encouraged as default validators in documentation." `strict_discrete_range` is very useful for instruments that require discrete step up/down in some range specification. Newer instruments can typically round to the correct discrete step, but often it is better to be explicit with what is sent to the instrument.

I don't think logging is a good solution to underlying issues, at least it doesn't solve the results file issue (unless you add a lot of plumbing for emitting the results), and as a log message it can be missed by the person running the experiment. 

This actually affected a researcher last week, who was using pymeasure on a hall effect characterization and accidentally set the Keithley 2400 source current to a current that was meant for programmable power supply. The `truncated_range` sent the truncated current value, which was enough to burn out the wirebonding and the device. This just caused wasted time and money on a relatively inexpensive device, but often such custom made devices can easily cost thousands of dollars. If it had been a `strict_range` validator it wouldn't have happened, so I'll be moving instruments over to `strict_*` validators. 

I know you can blame the user here, but never underestimate the power of defaults. That's why encouraging `strict_*` as default validators that can avoid a lot of hassles. I think instruments used in tutorial or used as a good implementation example should use the `strict_*` validators.

**Reply by bilderbuchi** [reply]
I think regarding the documentation, this can easily be changed. We're nearly there anyway (at least in the "adding instruments" file), most of the examples in there already use strict validators anyway, especially if you exclude those that specifically demonstrate the non-strict varieties.

> I don't think logging is a good solution to underlying issues, at least it doesn't solve the results file issue (unless you add a lot of plumbing for emitting the results), and as a log message it can be missed by the person running the experiment.

The results file issue is somewhat orthogonal to the logging issue, so that does not automatically discount the value of logging validation misses.
The project uses the logging framework throughout, we could maybe make this more prominent than [here](https://pymeasure.readthedocs.io/en/latest/tutorial/procedure.html#using-logs) so that users don't miss that so easily?

One idea I had on how we could "tighten the screws" if desired would be to have some global configuration (concrete mechanism TBD) where you could turn all pymeasure-provided validators into "strict" ones, if your setup necessitates that. My gut says that that should be achievable. This will not be a by-class or by-instance flag (validators are too "pure/functional" for that), but it could help in your situation. What do you think?

**Reply by BenediktBurger** [reply]
> One idea I had on how we could "tighten the screws" if desired would be to have some global configuration (concrete mechanism TBD) where you could turn all pymeasure-provided validators into "strict" ones, if your setup necessitates that. My gut says that that should be achievable. This will not be a by-class or by-instance flag (validators are too "pure/functional" for that), but it could help in your situation. What do you think?

Wouldn't it make sense, to unify both validators in one and have that flag determine the behavior?

**Reply by bilderbuchi** [reply]
The way we currently handle validators is that we expect a callable with 2 inputs, one output. If you want to add a flag to that you'd have to find a way to pass the desired flag as an additional argument when defining the property, e.g. by having the property's `validator` argument be `validator=partial(myvalidator, strict=True)`. I don't think that that is clearer than having two separate functions.

What _could_ make sense is to unify our implementations into one function (with appropriately defaulted `strict` kwarg), and have the _other_ function defined with `partial`. That's just for internal implementation, though, and could reduce code duplication.

**Reply by BenediktBurger** [reply]
I did not think about a case-by-case flag, but about that global flag you mentioned earlier, e.g.: `range_validator` will truncate if that global flag ist set to "truncate" and will raise an Exception, if the global flag is set to "raise".

**Reply by bilderbuchi** [reply]
Ah. In that case, all `range_validators` used anywhere have to use the same setting. I was thinking more of an optional way to get more strict behaviour if the situation demands it -- by default, instrument implementors should still be free to define strict/nonstrict behaviour as appropriate.

**Reply by mcdo0486** [reply]
> One idea I had on how we could "tighten the screws" if desired would be to have some global configuration (concrete mechanism TBD) where you could turn all pymeasure-provided validators into "strict" ones, if your setup necessitates that. My gut says that that should be achievable. This will not be a by-class or by-instance flag (validators are too "pure/functional" for that), but it could help in your situation. What do you think?

The current validators implementation is simple and functional which is a strength. I think adding global configuration flags can obfuscate behavior of the validators. For instance, if there was a global flag up the tree that had `VALIDATOR_BEHAVIOR = 'strict'` to switch the behavior of `valid_range_validator`, that wouldn't be obvious to the user when viewing an instrument property:
```
    source_current = Instrument.control(
        ":SOUR:CURR?", ":SOUR:CURR:LEV %g",
        """ A floating point property that controls the source current
        in Amps. """,
        validator=valid_range_validator,
        values=[-1.05, 1.05]
    )
```

I think refactoring the validator functions to accept `strict=True, **kwargs` would work well, `strict` can be passed as an optional argument in Instrument.control, which can be passed in `**kwargs` to the fset validator [here](https://github.com/pymeasure/pymeasure/blob/1751e6379e9bf7fbf1cdfc2b9808de8e3e039686/pymeasure/instruments/instrument.py#L467). That would make the strict behavior more obvious to the user: 

```
    source_current = Instrument.control(
        ":SOUR:CURR?", ":SOUR:CURR:LEV %g",
        """ A floating point property that controls the source current
        in Amps. """,
        validator=valid_range_validator
        values=[-1.05, 1.05],
        strict=False
    )
```

But we could just keep the validators logic simple, and recommend using `strict_*` and perhaps migrating validators that set instrument parameters over to `strict_*`.  `strict_*` is used by 62 instruments and `truncated_*` is used by 35 instruments, so I am not worried about migrating truncated to strict for the instruments I have access to.

**Reply by bilderbuchi** [reply]
> The current validators implementation is simple and functional which is a strength. 

Agreed.

> For instance, if there was a global flag up the tree that had VALIDATOR_BEHAVIOR = 'strict' to switch the behavior of valid_range_validator, that wouldn't be obvious to the user when viewing an instrument property:

It would be obvious to the user since the user has to explicitly _enable_ that behaviour (e.g. with a call like `pymeasure.force_strict_validators()` or somesuch).

> I think refactoring the validator functions to accept strict=True, **kwargs would work well, strict can be passed as an optional argument in Instrument.control, which can be passed in **kwargs to the fset validator [here](https://github.com/pymeasure/pymeasure/blob/1751e6379e9bf7fbf1cdfc2b9808de8e3e039686/pymeasure/instruments/instrument.py#L467).

How would you then adjust the behaviour of (existing) instruments you use, but want to show a more strict validator behaviour? Properties are created at import time. 
Instead of adding yet another arg to `control` (which I'm wary about, the signature is already quite large), we can just keep on using two separate validators, as we do now?

 >recommend using strict_* 

Agreed.

> and perhaps migrating validators that set instrument parameters over to strict_*.

Not so sure about that one, instrument implementor chose specific appropriate validators when they created the instruments, why change that? The change is not technically troublesome, but it will change behaviour for _existing_ code, which _is_ something that has to be considered well.

**Reply by BenediktBurger** [reply]
> Instead of adding yet another arg to control (which I'm wary about, the signature is already quite large), we can just keep on using two separate validators, as we do now?

I agree.

**Reply by mcdo0486** [reply]
> Not so sure about that one, instrument implementor chose specific appropriate validators when they created the instruments, why change that? The change is not technically troublesome, but it will change behaviour for _existing_ code, which _is_ something that has to be considered well.

My preference would be to keep the validator behavior as is, with simple validators that clearly denote what they do. In my own fork I'll be migrating validators over to `strict_`. Whether that should be pushed for pymeasure, as you say this can be classed as a backwards incompatible change to existing procedures. I don't have a strong opinion on that. However with this discussion and recent changes to "Adding Instruments" docs, maybe now is a good time to refactor instrument drivers. The number of drivers is only going to grow.

**Reply by bilderbuchi** [reply]
> My preference would be to keep the validator behavior as is, with simple validators that clearly denote what they do. In my own fork I'll be migrating validators over to strict_.

To make sure I understand correctly: you're saying you would not use/find useful a global switch to "strictify" all validators in a given experiment, and will rather maintain your own fork with all validators converted manually? I'm not judging -- it will be good to know so we can prioritze our limited manpower.

**Reply by bilderbuchi** [reply]
I'm tracking the documentation updates to emphasize strict validators in #729.
